cachi2/core/models/property_semantics.py

```diff
@@ -0,0 +1,89 @@
+import functools
+from dataclasses import dataclass, field
+from itertools import groupby
+from typing import TYPE_CHECKING, Iterable, Optional
+
+if TYPE_CHECKING:
+    from typing_extensions import Self, assert_never
+
+from cachi2.core.models.sbom import Component, Property
+
+
+def merge_component_properties(components: Iterable[Component]) -> list[Component]:
+    """Sort and de-duplicate components while merging their `properties`."""
+    components = sorted(components, key=Component.key)
+    grouped_components = groupby(components, key=Component.key)
+
+    def merge_component_group(component_group: Iterable[Component]) -> Component:
+        component_group = list(component_group)
+
+        prop_sets = (PropertySet.from_properties(c.properties) for c in component_group)
+        merged_prop_set = functools.reduce(PropertySet.merge, prop_sets)
+
+        hash_sets = (frozenset(c.hashes) for c in component_group)
+        merged_hash_set = functools.reduce(lambda a, b: a | b, hash_sets)
```

Comment:
I know this subverts the original intention of this module, but I think this is still the best place to merge the hashes. If you agree with it being here, I'll refactor this module into a generic "merge_components".

Reply:
I'm not sure if merging hashes is generically correct. In fact, I would say a hash mismatch should fail the request for every package manager except pip

Reply:
> a hash mismatch should fail the request for every package manager except pip

Actually, no. Yarnberry uses the checksum of the .zip archive in the cache, which means that:
- the same dependency will always have a different checksum between npm and Yarnberry
- the same dependency can have different checksums between two Yarnberry projects
  - if they use a different version of Yarnberry
  - or just a different compressionLevel setting

But anyway, let's worry about merging hashes when we work on adding hashes :smile:

Reply:
Yeah, the reasoning here was for a case in which a two different Pip projects had the same dependency, but with a different set of hashes. If we didn't merge hashes, only the ones in the first identified dependency would show up. But +1 on let's worry about this later, I already removed the code related to hashes :)